File: 0475-heaters/0475-heaters.cpp
class Solution {
public:
    bool isPossible(int radius,vector<int>& houses,vector<int>& heaters){
        int n=houses.size(),m=heaters.size(),i=0,j=0;
        while(i<n && j<m){
            int low=heaters[j]-radius,high=heaters[j]+radius;
            while(i<n && houses[i]>=low && houses[i]<=high){
                i++;
            }
            j++;
        }
        if(i==n){
            return true;
        }
        return false;
    }
    int findRadius(vector<int>& houses, vector<int>& heaters) {
        sort(houses.begin(),houses.end());
        sort(heaters.begin(),heaters.end());
        int i=0,j=1e9+1,ans=-1;
        while(i<=j){
            int mid=(i+j)/2;
            if(isPossible(mid,houses,heaters)){
                ans=mid;
                j=mid-1;
            }else{
                i=mid+1;
            }
        }
        return ans;
    }
};

File: 0475-heaters/0475-heaters.java
class Solution {
    public int findRadius(int[] houses, int[] heaters) {

        Arrays.sort(heaters);
        Arrays.sort(houses);
        
        int left = 0, right = 1000000001;
        int result = -1;
        while(left <= right){
            int mid = left + (right-left)/2;
            if(isPossible(mid, houses, heaters)){
                result = mid;
                if(result == 2){
                    System.out.println(left);
                    System.out.println(right);
                }
                right = mid-1;
            }
            else left = mid+1;
        }
        
        return result;
    }
    
    public boolean isPossible(int radius, int[] houses, int[] heaters){
        int ptr1 = 0, ptr2 = 0;
        
        while(ptr1 < houses.length && ptr2 < heaters.length){
            int low = heaters[ptr2] - radius, high = heaters[ptr2] + radius;
            while(ptr1 < houses.length && houses[ptr1] >= low && houses[ptr1] <= high){
                ptr1++;
            }
            ptr2++;
        }
        
        return ptr1 == houses.length;
    }
    
    
}

/*
[T,F,T,F,F,T]
[F,T,F,F,F,T]

[2]
 j

[1,5]
 j     
            r


[1,4] [1,7]

*/